feat: Дипломный проект. Этап 1

ChesnokovEK opened the pull request.

ChesnokovEK committed:
feat: Дипломный проект. Этап 1

KirPyt commented on core/main-service/src/main/java/ru/practicum/controller/admin/AdminCategoryController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @PatchMapping("/{catId}")
+    @ResponseStatus(HttpStatus.OK)
```
Данный статус идет по умолчанию, можно его не указывать

KirPyt commented on core/main-service/src/main/java/ru/practicum/controller/admin/AdminEventController.java:
```diff
@@ -0,0 +1,69 @@
+            @RequestParam(required = false) List<Long> categories,
+            @RequestParam(required = false) @DateTimeFormat(pattern = DATE_TIME_PATTERN) LocalDateTime rangeStart,
+            @RequestParam(required = false) @DateTimeFormat(pattern = DATE_TIME_PATTERN) LocalDateTime rangeEnd,
+            @RequestParam(defaultValue = "0") Integer from,
```
Рекомендую так же указать валидацию за счет `@Positive` и `@PositiveOrZero`

KirPyt commented on core/main-service/src/main/java/ru/practicum/controller/admin/AdminEventController.java:
```diff
@@ -0,0 +1,69 @@
+                users, states, categories, rangeStart, rangeEnd, from, size);
+        EventSearchParams eventSearchParams = new EventSearchParams();
+        AdminSearchParams adminSearchParams = new AdminSearchParams();
+        adminSearchParams.setUsers(users);
```
Рекомендую использовать конструктор, либо попробовать реализовать паттерн `Builder` (это можно сделать с применением lombok)

KirPyt commented on core/main-service/src/main/java/ru/practicum/controller/priv/PrivateLocationController.java:
```diff
@@ -0,0 +1,62 @@
+    @GetMapping("/top")
+    public List<LocationDto> getTop(
+            @PathVariable long userId,
+            @RequestParam(required = false, defaultValue = "10") Integer count,
```
При наличии `defaultValue` можно не писать `required = false`

KirPyt commented on core/main-service/src/main/java/ru/practicum/controller/publ/PublicEventController.java:
```diff
@@ -0,0 +1,135 @@
+            throw new IncorrectValueException("rangeStart of event can't be after rangeEnd");
+        }
+
+        EventSearchParams eventSearchParams = new EventSearchParams();
```
Так же можно попробовать описать мапперы, что отделить логику преобразования от контроллера

KirPyt commented on core/main-service/src/main/java/ru/practicum/entity/Event.java:
```diff
@@ -0,0 +1,74 @@
+    @Column(name = "EVENT_DATE")
+    private LocalDateTime eventDate;
+
+    @ManyToOne(fetch = FetchType.EAGER)
```
`FetchType.EAGER` для `ManyToOne` идет по умолчанию

KirPyt commented on core/main-service/src/main/java/ru/practicum/handler/ErrorHandler.java:
```diff
@@ -0,0 +1,93 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public ApiError handleException(final Exception e) {
```
Здесь стоит отлавливать более общее исключение - `Throwable`

KirPyt commented on core/main-service/src/main/java/ru/practicum/mapper/CategoryMapper.java:
```diff
@@ -0,0 +1,25 @@
+
+@UtilityClass
+public class CategoryMapper {
+    public static CategoryDto toCategoryDto(Category category) {
```
`static` при использовании `UtilityClass` можно не указывать

KirPyt commented on core/main-service/src/main/java/ru/practicum/service/LocationServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+        List<Location> locationTopList = locationRepository.findTop(count);
+
+        for (Location location : locationTopList) {
+            location.setLikes(locationRepository.countLikesByLocationId(location.getId()));
```
В целом, стоит постараться избегать циклических обращений к базе данных

KirPyt reviewed (commented):
Здравствуйте!

Указал комментарии по работе, в целом часть из комментариев может быть справедлива и для других подобных мест в коде, стоит перепроверить похожие места для учета комментариев. В целом так же рекомендую обратить внимание на обращения к сервисам или базе данных в цикле, это может привести к излишней загрузке запросами другого сервиса, по этой причине рекомендую подумать над тем, как можно оптимизировать подобные места в коде

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убраны дефолтные Http.ОК ответы
- Добавлены валидации @Positive, @PositiveOrZero
- Убраны required = false там где есть значения по умолчанию
- Реализован паттерн Builder

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- fix билдера

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- fix билдера

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Реализованы мапперы

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убраны fetch = FetchType.EAGER для @ManyToOne
- Изменен тип отлавливаемого исключения на Throwable
- Для UtilityClass убран static

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано циклическое обращение к БД в LocationService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Убрано одно из циклических обращений к БД в EventService

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- checkstyle

KirPyt commented on core/main-service/src/main/java/ru/practicum/controller/admin/AdminEventController.java:
```diff
@@ -43,22 +49,14 @@ public List<EventFullDto> getAll(
             @RequestParam(required = false) @DateTimeFormat(pattern = DATE_TIME_PATTERN) LocalDateTime rangeEnd,
-            @RequestParam(defaultValue = "0") Integer from,
-            @RequestParam(defaultValue = "10") Integer size
+            @RequestParam(defaultValue = "0") @PositiveOrZero Integer from,
```
Так же над классом надо разместить аннотацию `@Validated`, чтобы данные аннотации проверки сработали

KirPyt commented on core/main-service/src/main/java/ru/practicum/controller/params/mapper/AdminEventSearchParamsMapper.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.List;
+
+@Component
+public class AdminEventSearchParamsMapper {
```
Мапперы можно делать утилитарными, делая все методы в них статическими, чтобы не добавлять лишних бинов в контейнер

KirPyt reviewed (commented):
Здравствуйте!

Отлично, надеюсь комментарии по коду оказались полезными. Жду следующей части проекта!

Хорошего дня!

ChesnokovEK committed:
fix: Дипломный проект. Этап 1

- Изменил методы в мапперах на статичные
